File: nvidia/c-cpp/cusolver/examples/openacc_cusolver.cpp
#include <cuda_runtime.h>
#include <cusolverDn.h>

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <vector>

static void make_dense_system(int n, int nrhs, double *a, double *b) {
  for (int j = 0; j < n; ++j)
    for (int i = 0; i < n; ++i)
      a[i + (size_t)j * n] = i == j ? n + 1.0 : 1.0;
  for (int r = 0; r < nrhs; ++r)
    for (int i = 0; i < n; ++i)
      b[i + (size_t)r * n] = 2.0 * n;
}

int main() {
  const int n = 1024, nrhs = 16;
  std::vector<double> a((size_t)n * n), b((size_t)n * nrhs);
  std::vector<int> piv(n), getrf_info(1, -1), getrs_info(1, -1);
  make_dense_system(n, nrhs, a.data(), b.data());
  double *ap = a.data(), *bp = b.data();
  int *pp = piv.data(), *getrf_ip = getrf_info.data();
  int *getrs_ip = getrs_info.data();
  size_t ac = a.size(), bc = b.size();
  cusolverDnHandle_t handle = nullptr;
  if (cusolverDnCreate(&handle) != CUSOLVER_STATUS_SUCCESS) {
    std::fprintf(stderr, "cusolverDnCreate failed\n");
    return EXIT_FAILURE;
  }
  int lwork = 0;
  bool ok = true;
#pragma acc data copy(ap[0 : ac], bp[0 : bc]) create(pp[0 : n],              \
                                                     getrf_ip[0 : 1],         \
                                                     getrs_ip[0 : 1])
  {
#pragma acc host_data use_device(ap)
    {
      ok = cusolverDnDgetrf_bufferSize(handle, n, n, ap, n, &lwork) ==
           CUSOLVER_STATUS_SUCCESS;
    }
    double *device_work = nullptr;
    if (ok)
      ok = cudaMalloc((void **)&device_work, (size_t)lwork * sizeof(double)) ==
           cudaSuccess;
#pragma acc host_data use_device(ap, bp, pp, getrf_ip, getrs_ip)
    {
      if (ok)
        ok = cusolverDnDgetrf(handle, n, n, ap, n, device_work, pp,
                             getrf_ip) ==
                 CUSOLVER_STATUS_SUCCESS &&
             cusolverDnDgetrs(handle, CUBLAS_OP_N, n, nrhs, ap, n, pp, bp, n,
                              getrs_ip) == CUSOLVER_STATUS_SUCCESS;
    }
    ok = ok && cudaDeviceSynchronize() == cudaSuccess;
#pragma acc update self(getrf_ip[0 : 1], getrs_ip[0 : 1])
    if (device_work)
      cudaFree(device_work);
  }
  ok = ok && cusolverDnDestroy(handle) == CUSOLVER_STATUS_SUCCESS;
  double max_error = 0;
  for (double x : b)
    max_error = std::max(max_error, std::fabs(x - 1));
  std::printf("OpenACC-managed cuSOLVER solve complete; getrf info=%d, "
              "getrs info=%d, max error=%.17g\n",
              getrf_info[0], getrs_info[0], max_error);
  return ok && getrf_info[0] == 0 && getrs_info[0] == 0 &&
                 max_error <= 1e-12
             ? EXIT_SUCCESS
             : EXIT_FAILURE;
}

File: nvidia/fortran/cusolver/examples/openacc_cusolver.f90
program openacc_cusolver
    use cudafor
    use cusolverDn
    use cublas_v2, only: CUBLAS_OP_N
    use, intrinsic :: ieee_arithmetic, only: ieee_is_finite
    implicit none
    integer :: allocation_status

    interface
        subroutine make_dense_system(n, nrhs, mat_A, rhs_B)
            implicit none
            integer, intent(in) :: n, nrhs
            real(8), intent(out) :: mat_A(n,n), rhs_B(n,nrhs)
        end subroutine make_dense_system
    end interface

    integer, parameter :: n = 1024, nrhs = 16, lda = n, ldb = n
    integer :: istat, lwork, info_getrf, info_getrs
    type(cusolverDnHandle) :: handle
    real(8), allocatable :: mat_A(:,:), rhs_B(:,:)
    integer, allocatable :: ipiv(:)
    real(8), device, allocatable :: d_work(:)

    allocate(mat_A(n,n), rhs_B(n,nrhs), ipiv(n), stat=allocation_status)

    if (allocation_status /= 0) error stop 'allocation failed'
    call make_dense_system(n, nrhs, mat_A, rhs_B)
    istat = cusolverDnCreate(handle)
    if (istat /= 0) error stop 'cusolverDnCreate failed'

    !$acc data copyin(mat_A(1:n,1:n)) copy(rhs_B(1:n,1:nrhs)) &
    !$acc     create(ipiv(1:n)) copyout(info_getrf, info_getrs)
        !$acc host_data use_device(mat_A)
            istat = cusolverDnDgetrf_bufferSize( &
                handle, n, n, mat_A, lda, lwork)
    if (istat /= 0) error stop 'cusolverDnDgetrf_bufferSize failed'
        !$acc end host_data
        if (lwork < 0) error stop 'negative workspace size'
        allocate(d_work(max(1,lwork)), stat=allocation_status)
        if (allocation_status /= 0) error stop 'allocation failed'
        !$acc host_data use_device(mat_A, rhs_B, ipiv, &
        !$acc                      info_getrf, info_getrs)
            istat = cusolverDnDgetrf( &
                handle, n, n, mat_A, lda, d_work, ipiv, info_getrf)
    if (istat /= 0) error stop 'cusolverDnDgetrf failed'
        !$acc end host_data
        istat = cudaDeviceSynchronize()
        if (istat /= 0) error stop 'getrf synchronization failed'
        !$acc update self(info_getrf)
        if (info_getrf /= 0) error stop 'getrf failed before solve'
        !$acc host_data use_device(mat_A, rhs_B, ipiv, info_getrs)
            istat = cusolverDnDgetrs( &
                handle, CUBLAS_OP_N, n, nrhs, &
                mat_A, lda, ipiv, rhs_B, ldb, info_getrs)
    if (istat /= 0) error stop 'cusolverDnDgetrs failed'
        !$acc end host_data

        istat = cudaDeviceSynchronize()
    if (istat /= 0) error stop 'cudaDeviceSynchronize failed'
        deallocate(d_work)
    !$acc end data

    print *, "info_getrf = ", info_getrf, ", info_getrs = ", info_getrs
    print *, "rhs_B(1,1) = ", rhs_B(1,1)

    if (info_getrf /= 0 .or. info_getrs /= 0) error stop 'getrf/getrs failed'
    if (.not. all(ieee_is_finite(rhs_B))) error stop 'nonfinite solution'
    if (maxval(abs(rhs_B-1.0d0)) > 1.0d-10) error stop 'solution verification failed'
    block
        integer :: rhs
        real(8) :: residual, scale
        do rhs = 1, nrhs
            residual = maxval(abs(real(n,8)*rhs_B(:,rhs)+sum(rhs_B(:,rhs))-2.0d0*real(n,8)))
            scale = 2.0d0*real(n,8)*(maxval(abs(rhs_B(:,rhs)))+1.0d0)
            if (residual/scale > 1.0d-10) error stop 'residual verification failed'
        end do
    end block
    print *, 'verification PASS'
    istat = cusolverDnDestroy(handle)
    if (istat /= 0) error stop 'cusolverDnDestroy failed'
    deallocate(mat_A, rhs_B, ipiv)
end program openacc_cusolver
